cedar-policy-core/src/evaluator.rs

```diff
@@ -252,9 +252,383 @@ impl<'e> Evaluator<'e> {
+                        },
+                        BinaryOp::In,
+                    ) => Err(EvaluationError::type_error_with_advice(
+                        nonempty![Type::entity_type(names::ANY_ENTITY_TYPE.clone())],
+                        &arg1,
+                        "`in` is for checking the entity hierarchy;
+                        \
+                        use `.contains()` to test set membership"
+                            .into(),
+                    )),
+                    (
+                        _,
+                        Value {
+                            value: ValueKind::Record(_),
+                            ..
+                        },
+                        BinaryOp::In,
+                    ) => Err(EvaluationError::type_error_with_advice(
+                        nonempty![Type::entity_type(names::ANY_ENTITY_TYPE.clone())],
+                        &arg1,
+                        "`in` is for checking the entity hierarchy;
+                        \
+                        use `has` to test if a record has a key"
+                            .into(),
+                    )),
+                    (_, _, BinaryOp::In) => Err(EvaluationError::type_error_with_advice(
+                        nonempty![Type::entity_type(names::ANY_ENTITY_TYPE.clone())],
+                        &arg1,
+                        "the LHS of `in` should be an entity".into(),
+                    )),
+                    // contains, which works on Sets
+                    (_, _, BinaryOp::Contains) => match arg1.value {
+                        ValueKind::Set(Set { fast: Some(h), .. }) => Ok(arg2
+                            .try_as_lit()
+                            .map_or(false, |lit| h.contains(lit))
+                            .into()),
+                        ValueKind::Set(Set {
+                            fast: None,
+                            authoritative,
+                        }) => Ok((authoritative.contains(&arg2)).into()),
+                        _ => Err(EvaluationError::type_error_single(Type::Set, &arg1)),
+                    },
+                    // ContainsAll and ContainsAny, which work on Sets
+                    (
+                        Value {
+                            value:
+                                ValueKind::Set(Set {
+                                    authoritative: _,
+                                    fast: Some(arg1_set),
+                                }),
+                            ..
+                        },
+                        Value {
+                            value:
+                                ValueKind::Set(Set {
+                                    authoritative: _,
+                                    fast: Some(arg2_set),
+                                }),
+                            ..
+                        },
+                        BinaryOp::ContainsAll,
+                    ) => Ok((arg2_set.is_subset(&arg1_set)).into()),
```

Comment:
maybe would be good to copy the comment from the partial evaluator, that in this case both sets only contain literals and thus we have this fast hashset-based implementation